File: 213-house-robber-ii/house-robber-ii.cpp
class Solution {
public:
    int robRange(vector<int>& nums, int start, int end) {
        int prev = 0, curr = 0;
        for (int i = start; i <= end; ++i) {
            int next = max(curr, prev + nums[i]);
            prev = curr;
            curr = next;
        }
        return curr;
    }

    int rob(vector<int>& nums) {
        int N = nums.size();
        if (N == 0) return 0;
        if (N == 1) return nums[0];

        // Case 1: Rob houses 0 to N-2 (exclude the last house)
        int case1 = robRange(nums, 0, N - 2);
        
        // Case 2: Rob houses 1 to N-1 (exclude the first house)
        int case2 = robRange(nums, 1, N - 1);

        // Return the maximum of both cases
        return max(case1, case2);
    }
};

File: 213-house-robber-ii/house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;

        if(n == 0)return 0;
        if(n == 1)return nums[0];

        int case1 = roob(nums,0,n-2);
        int case2 = roob(nums,1,n-1);

        return Math.max(case1, case2);
    }
    public int roob(int [] nums, int start, int end){
        int prev=0; int curr=0;
        for(int i=start; i <= end; ++i){
            int next = Math.max(curr, prev+nums[i]);
            prev = curr;
            curr = next;
        }
        return curr;
    }
}


